### ISSUE TITLE
Assertion failure in ForwardingProcess::check_seq_ack()

### ISSUE THREAD

**Issue by kyechou** [netarch/neo#12]
```
./examples/04-squid/run.sh
```
Output: https://hastebin.com/roviyufeqe

**Comment by kyechou** [comment]
~Also, at line 84, the policy should not hold and should be violated because it was `PS_TCP_TERM_3` being delivered instead of `PS_TCP_L7_REQ`. Check `src/policy/reachability.cpp`.~

Edit; This should be a different issue.

**Comment by kyechou** [comment]
### Root cause

Simply replaying (resending) the history packets wouldn't work because the middlebox's raw seq number will be different for each new connection. During the attempt to rewind the middlebox state, the SYN-ACK packet from the middlebox will have a different raw seq number than the one in the packet history, and hence we need to reply the ACK packet (`PS_TCP_INIT_3`) with a different ack number (and also for the rest of the packets in the same connection).

**Comment by kyechou** [comment]
Relevant code location:
https://github.com/netarch/neo/blob/b0c9f7ccdd5df22a51a30190999a559913606caa/src/emulation.cpp#L161

Note that this is an issue only when a middlebox is an endpoint of a connection.

### Potential solution/comments
* Keep only the "relative" seq and ack numbers in the spin state.
* But how to remember the raw seq offset of the emulation endpoint with respect to the spin state during a DFS?
* Can we also abstract away the payload detail from the spin state?

**Comment by kyechou** [comment]
Similar to `Emulation::node_pkt_hist`, we should be able to have a variable within `class Emulation` to keep track of the seq and ack numbers offsets of the emulation when it is part of a communication endpoint. The variable will be reset whenever backtracking happens in the DFS, so that it doesn't have to be in the spin state.
```cpp
// conn -> { intf -> offset }
std::unordered_map<int, std::unordered_map<Interface *, uint32_t>> seq_ack_offsets;
```

Whenever a packet is injected to an emulation, we need to check the response packets in order to update the variable. (This may be done before L4 reassembly because we only need the SYN and SYN-ACK packets.)

If and when the response packet within a given connection remains in the same proto_state (phase), meaning that the middlebox is not an endpoint of that connection, we may just set the offset to be 0 or  make the entry non-existent.

Note that there can be multiple connections with the same interface as an endpoint, and there might be multiple emulated interface endpoints within the same connection (i.e., a connection consisting of 2 middleboxes).

**Comment by kyechou** [comment]
The issue should be resolved, but there's another similar issue (#15) with the source port numbers that needs to be fixed.
